components/gfx/display_list/mod.rs

```diff
@@ -521,26 +336,201 @@ impl DisplayList {
+                translated_point.x -= Au::from_f32_px(scroll_offset.x);
+                translated_point.y -= Au::from_f32_px(scroll_offset.y);
+            }
+        }
+
+        self.hit_test_contents(traversal, &translated_point, client_point, scroll_offsets, result);
+    }
+
+    pub fn print(&self) {
+        let mut print_tree = PrintTree::new("Display List".to_owned());
+        self.print_with_tree(&mut print_tree);
+    }
+
+    pub fn print_with_tree(&self, print_tree: &mut PrintTree) {
+        print_tree.new_level("Items".to_owned());
+        for item in &self.list {
+            print_tree.add_item(format!("{:?} StackingContext: {:?}",
+                                        item,
+                                        item.base().stacking_context_id));
+        }
+        print_tree.end_level();
+    }
+}
+
+pub struct DisplayListTraversal<'a> {
+    pub display_list: &'a DisplayList,
+    pub next_item_index: usize,
+    pub first_item_index: usize,
+    pub last_item_index: usize,
+}
+
+impl<'a> DisplayListTraversal<'a> {
+    pub fn new(display_list: &'a DisplayList) -> DisplayListTraversal {
+        DisplayListTraversal {
+            display_list: display_list,
+            next_item_index: 0,
+            first_item_index: 0,
+            last_item_index: display_list.list.len(),
+        }
+    }
+
+    pub fn new_partial(display_list: &'a DisplayList,
+                       stacking_context_id: StackingContextId,
+                       start: usize,
+                       end: usize)
+                       -> DisplayListTraversal {
+        assert!(display_list.list.len() > start);
+        assert!(display_list.list.len() > end);
+
+        let mut stacking_context_start = start;
+        for x in (0..start).rev()  {
+            if let DisplayItem::PushStackingContextClass(ref item) = display_list.list[x] {
+                if item.stacking_context.id == stacking_context_id {
+                    stacking_context_start = x;
+                    break;
+                }
+            }
+        }
+
+        assert!(stacking_context_start <= start);
+
+        DisplayListTraversal {
+            display_list: display_list,
+            next_item_index: stacking_context_start,
+            first_item_index: start,
+            last_item_index: end + 1,
+        }
+    }
+
+    pub fn previous_item_id(&self) -> usize {
+        return self.next_item_index - 1;
+    }
+
+    pub fn skip_to_end_of_stacking_context(&mut self, id: StackingContextId) {
+        loop {
+            assert!(self.next_item_index < self.last_item_index);
+
+            let item = &self.display_list.list[self.next_item_index];
+            if let &DisplayItem::PopStackingContextClass(ref item) = item {
+                if item.stacking_context_id == id {
+                    return;
+                }
+            }
+            self.next_item_index += 1;
+        }
```

Comment:
Maybe:

    self.next_item_index = self.display_list.list[self.next_item_index..].iter().position(|item| ...).unwrap_or(self.display_list.list.len())

Might be less readable though…up to you.